File: tests/commands/doctor.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { mkdtempSync, rmSync, mkdirSync, writeFileSync } from 'fs';
import { join } from 'path';
import { tmpdir } from 'os';
import { runDoctor } from '../../src/commands/doctor.js';

const goodSkill = `
skill: good
version: "1.0.0"
context:
  consumes: []
  produces: []
  memory: short-term
strategy:
  tools: [read_file]
  approach: sequential
guardrails:
  - max_tokens: 1000
depends_on: []
observability:
  trace_level: standard
  metrics: [tokens]
security:
  filesystem: read-only
  network: none
  secrets: []
negotiation:
  file_conflicts: yield
  priority: 0
`;

const dangerousSkill = `
skill: dangerous
version: "1.0.0"
context:
  consumes: [data]
  produces: [data]
  memory: long-term
strategy:
  tools: []
  approach: sequential
guardrails: []
depends_on:
  - skill: nonexistent
    provides: something
observability:
  trace_level: minimal
  metrics: []
security:
  filesystem: full
  network: none
  secrets: []
negotiation:
  file_conflicts: yield
  priority: 0
`;

describe('forgent doctor command', () => {
  let tempDir: string;

  beforeEach(() => {
    tempDir = mkdtempSync(join(tmpdir(), 'forgent-doctor-'));
    mkdirSync(join(tempDir, 'skills'));
  });

  afterEach(() => {
    rmSync(tempDir, { recursive: true, force: true });
  });

  it('should report high score for well-configured skills', () => {
    writeFileSync(join(tempDir, 'skills', 'good.skill.yaml'), goodSkill);
    const report = runDoctor(join(tempDir, 'skills'));
    expect(report.skills).toHaveLength(1);
    expect(report.parseErrors).toHaveLength(0);
    expect(report.dependencyIssues).toHaveLength(0);
    expect(report.score).toBeGreaterThanOrEqual(80);
  });

  it('should detect all issues in a dangerous skill', () => {
    writeFileSync(join(tempDir, 'skills', 'dangerous.skill.yaml'), dangerousSkill);
    const report = runDoctor(join(tempDir, 'skills'));
    expect(report.skills).toHaveLength(1);
    expect(report.dependencyIssues.length).toBeGreaterThan(0);
    expect(report.loopRisks.size).toBeGreaterThan(0);
    expect(report.lintIssues.size).toBeGreaterThan(0);
  });

  it('should handle parse errors gracefully', () => {
    writeFileSync(join(tempDir, 'skills', 'broken.skill.yaml'), 'invalid: yaml: [[[');
    const report = runDoctor(join(tempDir, 'skills'));
    expect(report.parseErrors).toHaveLength(1);
    expect(report.skills).toHaveLength(0);
  });

  it('should aggregate issues from multiple skills', () => {
    writeFileSync(join(tempDir, 'skills', 'good.skill.yaml'), goodSkill);
    writeFileSync(join(tempDir, 'skills', 'dangerous.skill.yaml'), dangerousSkill);
    const report = runDoctor(join(tempDir, 'skills'));
    expect(report.skills).toHaveLength(2);
  });

  it('should return score of 100 for empty directory', () => {
    const report = runDoctor(join(tempDir, 'skills'));
    expect(report.skills).toHaveLength(0);
    expect(report.score).toBe(100);
  });

  it('should detect skill ordering issues when agentsDir is provided', () => {
    const consumerSkill = `
skill: consumer
version: "1.0.0"
context:
  consumes: [data_from_producer]
  produces: [result]
  memory: short-term
strategy:
  tools: [read_file]
  approach: sequential
guardrails:
  - timeout: 5min
depends_on: []
observability:
  trace_level: minimal
  metrics: []
security:
  filesystem: read-only
  network: none
  secrets: []
negotiation:
  file_conflicts: yield
  priority: 0
`;
    const producerSkill = `
skill: producer
version: "1.0.0"
context:
  consumes: []
  produces: [data_from_producer]
  memory: short-term
strategy:
  tools: [read_file]
  approach: sequential
guardrails:
  - timeout: 5min
depends_on: []
observability:
  trace_level: minimal
  metrics: []
security:
  filesystem: read-only
  network: none
  secrets: []
negotiation:
  file_conflicts: yield
  priority: 0
`;
    const badOrderAgent = `
agent: pipeline
skills: [consumer, producer]
orchestration: sequential
description: "Consumer runs before producer"
`;
    writeFileSync(join(tempDir, 'skills', 'consumer.skill.yaml'), consumerSkill);
    writeFileSync(join(tempDir, 'skills', 'producer.skill.yaml'), producerSkill);
    mkdirSync(join(tempDir, 'agents'));
    writeFileSync(join(tempDir, 'agents', 'pipeline.agent.yaml'), badOrderAgent);
    const report = runDoctor(join(tempDir, 'skills'), join(tempDir, 'agents'));
    expect(report.orderingIssues.length).toBeGreaterThan(0);
    expect(report.orderingIssues[0].type).toBe('data-flow-order-mismatch');
  });

  it('should work without agentsDir parameter', () => {
    writeFileSync(join(tempDir, 'skills', 'good.skill.yaml'), goodSkill);
    const report = runDoctor(join(tempDir, 'skills'));
    expect(report.orderingIssues).toHaveLength(0);
  });
});

File: internal/cmd/doctor_test.go
package cmd

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

const doctorValidSkillYAML = `skill: healthy-skill
version: "0.1.0"
context:
  consumes: []
  produces: []
  memory: short-term
strategy:
  tools:
    - Read
  approach: sequential
guardrails:
  - "timeout: 30s"
depends_on: []
observability:
  trace_level: minimal
  metrics:
    - latency
security:
  filesystem: none
  network: none
  secrets: []
negotiation:
  file_conflicts: yield
  priority: 0
`

const doctorInvalidSkillYAML = `skill: ""
version: ""
`

const doctorLintIssuesSkillYAML = `skill: risky-skill
version: "0.1.0"
context:
  consumes: []
  produces:
    - output
  memory: short-term
strategy:
  tools: []
  approach: sequential
guardrails: []
depends_on: []
observability:
  trace_level: minimal
  metrics: []
security:
  filesystem: read-write
  network: none
  secrets: []
negotiation:
  file_conflicts: yield
  priority: 0
`

func TestRunDoctor_ValidSkills_HighScore(t *testing.T) {
	dir := t.TempDir()
	err := os.WriteFile(filepath.Join(dir, "healthy.skill.yaml"), []byte(doctorValidSkillYAML), 0644)
	require.NoError(t, err)

	report := RunDoctor(dir, "")

	assert.Len(t, report.Skills, 1)
	assert.Empty(t, report.ParseErrors)
	assert.Empty(t, report.DependencyIssues)
	assert.Equal(t, 100, report.Score)
}

func TestRunDoctor_InvalidSkill_ParseErrors(t *testing.T) {
	dir := t.TempDir()
	err := os.WriteFile(filepath.Join(dir, "bad.skill.yaml"), []byte(doctorInvalidSkillYAML), 0644)
	require.NoError(t, err)

	report := RunDoctor(dir, "")

	assert.Empty(t, report.Skills)
	assert.Len(t, report.ParseErrors, 1)
	assert.Equal(t, "bad.skill.yaml", report.ParseErrors[0].File)
	assert.Less(t, report.Score, 100)
}

func TestRunDoctor_EmptyDirectory_ZeroSkills(t *testing.T) {
	dir := t.TempDir()

	report := RunDoctor(dir, "")

	assert.Empty(t, report.Skills)
	assert.Empty(t, report.ParseErrors)
	assert.Equal(t, 100, report.Score) // No issues = perfect score
}

func TestRunDoctor_LintIssues_Populated(t *testing.T) {
	dir := t.TempDir()
	err := os.WriteFile(filepath.Join(dir, "risky.skill.yaml"), []byte(doctorLintIssuesSkillYAML), 0644)
	require.NoError(t, err)

	report := RunDoctor(dir, "")

	assert.Len(t, report.Skills, 1)
	assert.Empty(t, report.ParseErrors)

	issues, ok := report.LintIssues["risky-skill"]
	assert.True(t, ok, "expected lint issues for risky-skill")
	assert.Greater(t, len(issues), 0)
}

func TestRunDoctor_NonExistentDirectory(t *testing.T) {
	report := RunDoctor("/nonexistent/path", "")

	assert.Empty(t, report.Skills)
	assert.Empty(t, report.ParseErrors)
	assert.Equal(t, 100, report.Score)
}

func TestRunDoctor_WithAgents_OrderingIssues(t *testing.T) {
	skillsDir := t.TempDir()
	agentsDir := t.TempDir()

	// Create two skills with data flow
	skill1 := `skill: producer
version: "0.1.0"
context:
  consumes: []
  produces:
    - data
  memory: short-term
strategy:
  tools:
    - Read
  approach: sequential
guardrails:
  - "timeout: 30s"
depends_on: []
observability:
  trace_level: minimal
  metrics:
    - latency
security:
  filesystem: none
  network: none
  secrets: []
negotiation:
  file_conflicts: yield
  priority: 0
`
	skill2 := `skill: consumer
version: "0.1.0"
context:
  consumes:
    - data
  produces: []
  memory: short-term
strategy:
  tools:
    - Read
  approach: sequential
guardrails:
  - "timeout: 30s"
depends_on: []
observability:
  trace_level: minimal
  metrics:
    - latency
security:
  filesystem: none
  network: none
  secrets: []
negotiation:
  file_conflicts: yield
  priority: 0
`
	// Agent with wrong ordering: consumer before producer
	agentYAML := `agent: bad-pipeline
skills:
  - consumer
  - producer
orchestration: sequential
description: "A pipeline that runs consumer then producer"
`

	require.NoError(t, os.WriteFile(filepath.Join(skillsDir, "producer.skill.yaml"), []byte(skill1), 0644))
	require.NoError(t, os.WriteFile(filepath.Join(skillsDir, "consumer.skill.yaml"), []byte(skill2), 0644))
	require.NoError(t, os.WriteFile(filepath.Join(agentsDir, "bad-pipeline.agent.yaml"), []byte(agentYAML), 0644))

	report := RunDoctor(skillsDir, agentsDir)

	assert.Len(t, report.Skills, 2)
	assert.Greater(t, len(report.OrderingIssues), 0)
}
